torch/distributed/tensor/parallel/style.py

```diff
@@ -242,6 +243,82 @@ def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
             partial(self._prepare_output_fn, self.output_layouts, self.use_local_output),
         )
 
+class SequenceParallel(ParallelStyle):
+    """
+    SequenceParallel replicate a compatible nn.Module parameters and run the sharded computation with the sharded
+    input that shards on the sequence dimension. Currently supporting ``nn.LayerNorm``, ``nn.Dropout`` and the
+    `RMSNorm python implementation <https://github.com/facebookresearch/llama/blob/main/llama/model.py#L34>`__
+
+    This style implements the operation that described in the paper
+    `Reducing Activation Recomputation in Large Transformer Models <https://arxiv.org/abs/2205.05198>`__
+
+    Both the input and output of the nn.Module will be sharded on the sequence dimension.
+
+    Keyword Args:
+        sequence_dim (int, optional):
+            The sequence dimension of the input tensor for the nn.Module, this is used to annotate the input tensor to
+            become a DTensor that is sharded on the sequence dimension.
+        use_local_output (bool, optional):
+            Whether to use local :class:`torch.Tensor` instead of :class:`DTensor` for the module output, default: False.
+    Returns:
+        A :class:`ParallelStyle` object that represents Sequence Parallel of the nn.Module.
+
+    Example::
+        >>> # xdoctest: +SKIP(failing)
+        >>> from torch.distributed.tensor.parallel import parallelize_module, SequenceParallel
+        >>> from torch.distributed.device_mesh import init_device_mesh
+        >>> ...
+        >>> m = Model(...)  # m is a nn.Module that contains a "norm" nn.LayerNorm submodule
+        >>> tp_mesh = init_device_mesh("cuda", (8,))
+        >>>
+        >>> # By default, the input of the "norm" will be converted to DTensor that shards on the sequence dim
+        >>> # and the output of "norm" will return a sharded on sequence dimension :class:`torch.Tensor`.
+        >>>
+        >>> sharded_mod = parallelize_module(m, tp_mesh, {"norm": SequenceParallel()}),
+        >>> ...
+    """
+    def __init__(
+        self,
+        *,
+        sequence_dim: int = 1,
+        use_local_output: bool = False
+    ):
+        super().__init__()
+        self.sequence_dim = sequence_dim
+        self.use_local_output = use_local_output
+
+    def _replicate_module_fn(self, name: str, module: nn.Module, device_mesh: DeviceMesh):
+        for p_name, param in module.named_parameters():
+            # simple replication with fixed ones_ init from LayerNorm/RMSNorm, which allow
```

Comment:
Would it be safer to include a broadcast? I worry that the failure mode of this assumption would be divergent model replicas, which might be hard to debug.

Reply:
So for all our "claimed" supporting modules, this is not a issue:
* nn.LayerNorm, after init it always have weights are that ones, https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/normalization.py#L192
* similarly RMSNormPython also have ones weight
* nn.Dropout does not have parameters

So I think it's very safe to just mark the tensor as Replicate without doing communication during init, let me know if this make sense

Reply:
added a docstr to explicitly mention the `ones` init assumption